### PR TITLE
Add RHOS support

### LINKED ISSUE #1725: Enhance subctl to support cloud prepare for RHOS
**What would you like to be added**:
Enhance subctl to support cloud prepare for RHOS

**Why is this needed**:
This is needed support submariner in RHOS deployment.

### PR DESCRIPTION
**Pull request by aswinsuryan**
Depends on submariner-io/cloud-prepare#188

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

### CODE REVIEW

**Comment by aswinsuryan** [comment]
Depends on -> https://github.com/submariner-io/cloud-prepare/pull/188/

**Comment by aswinsuryan** [comment]
The test will pass only after submariner-io/cloud-prepare#188 is merged and dependency is updated.

**Comment by nyechiel** [comment]
Depends on https://github.com/submariner-io/cloud-prepare/pull/188/

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+package cleanup
+
+import (
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/rhos"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/utils"
+)
+
+// NewCommand returns a new cobra.Command used to prepare a cloud infrastructure.
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go]:
```suggestion
// newRHOSCleanupCommand returns a new cobra.Command used to cleanup a cloud infrastructure.
```

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+func newRHOSCleanupCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "rhos",
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go:31]: Just a note (no action required in this pr): We are using generic names like rhos, aws, gcp etc but they work mainly with OCP alone. Probably we should use names to convey that its for OCP on aws/gcp/rhos etc.

File: pkg/subctl/cmd/cloud/prepare/rhos.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/rhos"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/utils"
+)
+
+// NewCommand returns a new cobra.Command used to prepare a cloud infrastructure.
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/cloud/prepare/rhos.go]:
```suggestion
// newRHOSPrepareCommand returns a new cobra.Command used to prepare a cloud infrastructure.
```

File: pkg/subctl/cmd/cloud/rhos/rhos.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+	cloudInfo := rhos.CloudInfo{
+		Client:    providerClient,
+		InfraID:   infraID,
+		Region:    region,
+		K8sClient: k8sClientSet,
+	}
+	rhosCloud := rhos.NewCloud(cloudInfo)
+	// TODO: Ideally we should be able to specify the image for GWNode, but it was seen that
+	// with certain images, the instance is not coming up. Needs to be investigated further.
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/cloud/rhos/rhos.go]:
```suggestion

```

I think this is not applicable to this implementation

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Some minor comments, otherwise looks good.

**Review by sridhargaddam — APPROVED** [review]

File: pkg/subctl/cmd/cloud/cleanup/rhos.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+		func(cloud api.Cloud, gwDeployer api.GatewayDeployer, reporter api.Reporter) error {
+			err := gwDeployer.Cleanup(reporter)
+			if err != nil {
+				return err
+			}
+
+			return cloud.CleanupAfterSubmariner(reporter)
+		})
+
+	utils.ExitOnError("Failed to cleanup RHOS cloud", err)
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/rhos.go]: Let's use the new API in _github.com/submariner-io/submariner-operator/internal/exit_. Please change all other places as well.

```suggestion
	exit.OnErrorWithMessage("Failed to cleanup RHOS cloud", err)
```

File: pkg/subctl/cmd/cloud/rhos/rhos.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	opts := &clientconfig.ClientOpts{
+		Cloud: cloudEntry,
+	}
+
+	providerClient, err := clientconfig.AuthenticatedClient(opts)
+	if err != nil {
+		panic(err)
+	}
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/rhos/rhos.go]:
```suggestion
```

File: pkg/subctl/cmd/cloud/rhos/rhos.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+// AddRHOSFlags adds basic flags needed by RHOS.
+func AddRHOSFlags(command *cobra.Command) {
+	command.Flags().StringVar(&infraID, infraIDFlag, "", "RHOS infra ID")
+	command.Flags().StringVar(&region, regionFlag, "", "RHOS region")
+	command.Flags().StringVar(&projectID, projectIDFlag, "", "RHOS project ID")
+	command.Flags().StringVar(&ocpMetadataFile, "ocp-metadata", "",
+		"OCP metadata.json file (or the directory containing it) from which to read the RHOS infra ID "+
+			"and region from (takes precedence over the specific flags)")
+	command.Flags().StringVar(&cloudEntry, cloudEntryFlag, "", " the cloud entry to use")
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/rhos/rhos.go]:
```suggestion
	command.Flags().StringVar(&cloudEntry, cloudEntryFlag, "", "the cloud entry to use")
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Pull request merged**